### DISCUSSION TITLE
FZF_CTRL_T_OPTS evaluates IF statement with system default shell

### DISCUSSION THREAD

**Discussion by lukejanicke** [junegunn/fzf#4108] [Q&A] [answered]
I switch between Zsh and Fish on Mac.

I have found that FZF evaluates IF statements with the system default shell and not the current shell, if it is different. I am using an IF statement to conditionally set `eza` or `bat` in FZF_CTRL_T_OPTS. As a result, I can only use FZF like this when my current and system default shell match. I would like to better understand why and maybe resolve this limitation. My knowledge of shells is limited but I have done extensive troubleshooting to get to this point of understanding the situation. Hope someone can help!

Here are my Zsh and Fish configurations:

```shell
# ZSH ~/.zshrc
# CONTROL + T to preview file content using bat or directory structure using eza
export FZF_CTRL_T_OPTS="
  --walker-skip .git,node_modules,target
  --preview 'if [ -d {} ]; then eza --tree --color=always {}; else bat -n --color=always {}; fi'
  --bind 'ctrl-/:change-preview-window(down|hidden|)'
  --color header:italic
  --header 'Use CONTROL + / to toggle preview window'"
```
  
```shell
# FISH ~/.config/fish/config.fish
# CONTROL + T to preview file content using bat or directory structure using eza
set -x FZF_CTRL_T_OPTS "
  --walker-skip .git,node_modules,target
  --preview 'if test -d {}; eza --tree --colour=always --icons=always {}; else; bat -n --color=always {}; end'
  --bind 'ctrl-/:change-preview-window(down|hidden|)'
  --color header:italic
  --header 'Use CONTROL + / to toggle preview window'"
```

When I am in Zsh, but the system default shell is Fish, the IF statement in the Zsh configuration causes an error that is reported as a Fish error. And vice versa. When I am in Fish, but the system default shell is Zsh, the IF statement in the Fish configuration causes an error that is reported as a Zsh error.

In this screenshot, I have just set the system default shell to `fish` with `sudo dscl . -create /Users/$(whoami) UserShell /opt/homebrew/bin/fish` and opened a new terminal. I then switch to `zsh` and press CONTROL + Y. Here, you can clearly see that `fish` is reporting an error but showing the IF string from my `zsh` configuration. This means the IF string is being evaluated with a `fish` shell (the current system default) even though I am now in `zsh`.

<img width="1566" alt="Screenshot 2024-11-22 at 10 41 10 pm" src="https://github.com/user-attachments/assets/87eda600-5ed5-44b9-93cd-5be6dfe3e55a">

In this screenshot, I have just set the system default shell to `zsh` with `sudo dscl . -create /Users/$(whoami) UserShell /bin/zsh` and opened a new terminal. I then switch to `fish` and press CONTROL + Y. The error is reported as a `zsh` error, even though I am now in `fish`. With some messing around, I was able to confirm that `zsh` is indeed getting the IF string from my `fish` configuration.

<img width="1566" alt="Screenshot 2024-11-22 at 10 39 21 pm" src="https://github.com/user-attachments/assets/91a5015a-6391-490b-b6e2-c4cdc4c031e9">

Here is a screenshot where it works because I am in `fish` and the system default shell is also `fish`.

<img width="1566" alt="Screenshot 2024-11-22 at 10 41 45 pm" src="https://github.com/user-attachments/assets/9924b086-dbf5-4b6e-9a03-f3f63156085d">

**Comment by LangLangBart** [comment, marked as answer]
> I have found that FZF evaluates IF statements with the system default shell and not the current shell, if it is different.

> [!NOTE]
> _fzf runs the command with `$SHELL -c` if `SHELL` is set, otherwise with `sh -c`_

Try assigning `SHELL` in your `zsh` or `fish` config file to see if it works for you.
```bash
fzf --bind 'start:become:echo zsh $ZSH_VERSION fish $FISH_VERSION bash $BASH_VERSION'
# zsh 5.9 fish bash

SHELL=$(which fish)

fzf --bind 'start:become:echo zsh $ZSH_VERSION fish $FISH_VERSION bash $BASH_VERSION'
# zsh fish 3.7.1 bash
```

Alternatively, you can pass the `--with-shell=STR`[^1] flag to `fzf`.
```bash
fzf --with-shell="$(which zsh) -c" \
  --bind 'start:become:echo zsh $ZSH_VERSION fish $FISH_VERSION bash $BASH_VERSION'
# zsh 5.9 fish bash

fzf --with-shell="$(which fish) -c" \
  --bind 'start:become:echo zsh $ZSH_VERSION fish $FISH_VERSION bash $BASH_VERSION'
# zsh fish 3.7.1 bash

fzf --with-shell="$(which bash) -c" \
  --bind 'start:become:echo zsh $ZSH_VERSION fish $FISH_VERSION bash $BASH_VERSION'
# zsh fish bash 5.2.37(1)-release
```

[^1]: [fzf/CHANGELOG.md 0.51.0 ](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0510)

**Reply by lukejanicke** [reply]
Thank you! That works! ❤️

```shell
# ~/.config/fish/config.fish

# Set SHELL for Fish sessions
set -x SHELL (which fish)
```

```shell
# ~/.zshrc

# Set SHELL for Zsh sessions
export SHELL=$(which zsh)
```

And I’ll keep the `--with-shell` approach up my sleeve too! Much appreciated!
